src/udev/net/ethtool-util.c

```diff
@@ -473,15 +473,16 @@ static int set_slinksettings(int *fd, struct ifreq *ifr, const struct ethtool_li
         struct {
                 struct ethtool_link_settings req;
                 __u32 link_mode_data[3 * ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32];
-        } ecmd = {
-                .req.cmd = ETHTOOL_SLINKSETTINGS,
-        };
+        } ecmd;
         unsigned int offset;
         int r;
 
         if (u->base.cmd != ETHTOOL_GLINKSETTINGS || u->base.link_mode_masks_nwords <= 0)
                 return -EINVAL;
 
+        memset(&ecmd, sizeof(ecmd), 0);
```

Comment:
It is not necessary to clear whole area. Just `link_mode_data` is sufficient, right?

Reply:
It may be still overkill. I do not know whether the space needs to clear or not, but if it is necessary, then clear the rest of space after `memcpy(&ecmd.link_mode_data[offset], u->link_modes.lp_advertising, 4 * ecmd.req.link_mode_masks_nwords);`.

Reply:
My rationale was that the previous code seemed to be zeroing the entire block:
```
static int set_slinksettings(int *fd, struct ifreq *ifr, const struct ethtool_link_usettings *u) {
        struct {
                struct ethtool_link_settings req;
                __u32 link_mode_data[3 * ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32];
        } ecmd = {
                .req.cmd = ETHTOOL_SLINKSETTINGS,
        };
```
Without that initialization, I believe _ecmd_ comes into being with whatever's in the stack and would need to be zeroed. On my target, the function is being executed once for each link -- when the device is enabled. I haven't looked at assembly instructions generated by a compiler in over 20 years, and I'm not that good a coder, so feel free to make any optimizations you think necessary. The critical thing is that the base parameters weren't being copied, and they need to be. Thank you for your help!

Reply:
Please initialize ecmd to zero during declaration, the compiler usually generates better code then. Hence, please change the declaration of the variable to `struct { … } ecmd = {};` and drop the `memset()` here

Reply:
This is pushed in ac386980. Thanks!